File: May-LeetCoding-Challenge/22-Sort-Characters-By-Frequency/Sort-Characters-By-Frequency.java
class Solution {
    public String frequencySort(String s) {
        StringBuilder S = new StringBuilder();
        int table[] = new int[128];
        for(int i=0; i<s.length(); ++i){
            ++table[s.charAt(i)];
        }
        
        int max = 0;
        int temp = table[max];
        do {
            max = 0;
            temp = table[max];
            for(int j=0; j<128; ++j){
                if(table[j]> temp){
                    max = j;
                    temp = table[max];
                }
            }
            for(int k=temp; k>0; --k){
                S.append((char)max);
            }
            table[max] = 0;
        } while(temp != 0);
        return S.toString();
    }
}

File: May-LeetCoding-Challenge/22-Sort-Characters-By-Frequency/Sort-Characters-By-Frequency.kt
class SortCharactersByFrequencyKotlin451 {
    fun frequencySort(s: String): String {
        val charCountsMap: MutableMap<Char, Int> = HashMap()
        s.forEach {
            charCountsMap[it] = 1 + charCountsMap.getOrDefault(it, 0)
        }
        val result = StringBuilder(s.length)
        charCountsMap.toList()
            .sortedByDescending { it.second }
            .forEach {
                result.append(it.first.toString().repeat(it.second))
            }
        return result.toString()
    }
    /*
    fun frequencySort(s: String): String {
        val charCountsMap: MutableMap<Char, Int> = HashMap()
        s.forEach {
            charCountsMap[it] = 1 + charCountsMap.getOrDefault(it, 0)
        }
        return charCountsMap.toList()
            .sortedByDescending { it.second }.joinToString(separator = "") { it.first.toString().repeat(it.second) }
    }
     */
}

fun main() {
    val solution = SortCharactersByFrequencyKotlin451()
    println(solution.frequencySort("tree"))
    println(solution.frequencySort("cccaaa"))
}